signer/src/storage/memory/store.rs

```diff
@@ -0,0 +1,415 @@
+        .max_by_key(|block| (block.block_height, block.block_hash.to_bytes()))
+        .cloned()
+    }
+
+    pub(super) fn get_withdrawal_requests(
+        &self,
+        chain_tip: &model::BitcoinBlockHash,
+        context_window: u16,
+    ) -> Vec<model::WithdrawalRequest> {
+        let first_block = self.bitcoin_blocks.get(chain_tip);
+
+        let context_window_end_block = std::iter::successors(first_block, |block| {
+            Some(self.bitcoin_blocks.get(&block.parent_hash).unwrap_or(block))
+        })
+        .nth(context_window as usize);
+
+        let Some(context_window_end_block) = context_window_end_block else {
+            return Vec::new();
+        };
+
+        let Some(stacks_chain_tip) = self.get_stacks_chain_tip(chain_tip) else {
+            return Vec::new();
+        };
+
+        std::iter::successors(Some(&stacks_chain_tip), |stacks_block| {
+            self.stacks_blocks.get(&stacks_block.parent_hash)
+        })
+        .take_while(|stacks_block| {
+            self.bitcoin_blocks
+                .get(&stacks_block.bitcoin_anchor)
+                .is_some_and(|anchor| anchor.block_height >= context_window_end_block.block_height)
+        })
+        .flat_map(|stacks_block| {
+            self.stacks_block_to_withdrawal_requests
+                .get(&stacks_block.block_hash)
+                .cloned()
+                .unwrap_or_default()
+                .into_iter()
+                .map(|pk| {
+                    self.withdrawal_requests
+                        .get(&pk)
+                        .expect("missing withdraw request")
+                        .clone()
+                })
+        })
+        .collect()
+    }
+}
+
+impl Transactable for SharedStore {
+    type Tx<'a> = InMemoryTransaction;
+
+    async fn begin_transaction(&self) -> Result<Self::Tx<'_>, Error> {
+        let store = self.lock().await;
+        let store_clone = store.clone();
+        Ok(InMemoryTransaction {
+            version: store.version,
+            store: Arc::new(Mutex::new(store_clone)),
+            original_store_mutex: Arc::clone(self),
+            completed: AtomicBool::new(false),
+        })
+    }
+}
+
+/// Represents an active in-memory transaction.
+pub struct InMemoryTransaction {
+    /// Records the version of the store at the time of transaction creation.
+    pub version: usize,
+    /// Holds a clone of the store's data for operations within this transaction.
+    pub store: SharedStore,
+    /// Reference to the original store's mutex to commit changes back.
+    pub original_store_mutex: SharedStore,
+    /// Tracks if commit/rollback has been called.
+    pub completed: AtomicBool,
+}
+
+impl TransactionHandle for InMemoryTransaction {
+    async fn commit(self) -> Result<(), Error> {
+        if self.completed.load(std::sync::atomic::Ordering::SeqCst) {
+            panic!("Transaction already completed");
```

Comment:
Why panicking, vs returning an error? also below

Reply:
I was just lazy and didn't want to add a bunch of additional error variants for the in-mem db. Maybe I should just add a single `InMemoryDb(&'static str)` variant and replace the one I added for optimistic concurrency...

Reply:
It's test code so no biggie, OTOH it does seem one we could avoid (given below we use an error). I would prefer removing it, but no strong opinions either way, up to you (feel free to resolve this).

Reply:
I agree, I'm going to do like I suggested above with just a single error variant that takes a string, so that we're at least returning an error. It kinda sucks panicking because it makes it more difficult to see what call actually caused the panic in tests, so better to unwrap in the test and get the actual callsite.

Reply:
[82ef200](https://github.com/stacks-sbtc/sbtc/pull/1674/commits/82ef2004c9c8ecba7910bc4bf888a276694a2fa2) + [1288a72](https://github.com/stacks-sbtc/sbtc/pull/1674/commits/1288a72ea5bc52f8e9b46a8383cbbaaf983cbe5d)